---
title: "Genome Informatics (Part I)"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Section 1: Identify Genetic Variants of Interest

## 1000 Genomes Data Analysis 

Downloaded genotype data from Ensemble for Asthma associated SNP rs8067378
See: (MXL) Mexican Ancestry in Los Angeles, California 

```{r}
# Read CSV file
genotype <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")

head(genotype)
```

Find how many of each genotype there are in this dataset.

```{r}
genotype.summary <- table(genotype[,2])
```

Proportion of each genotype

```{r}
genotype.summary/nrow(genotype) * 100
```

**Q5.** What proportion of the Mexican Ancestry in Los Angeles sample population (MXL)
are homozygous for the asthma associated SNP (G|G)?

**A5.** 14%

# Section 2: Initial RNA-Seq Analysis 

## RNA-Seq
Look more closely at sample HG00109 with G|G gentoype for this SNP (from GBR population). 

### Side note on Fastq quality scores 
```{r}
# install.packages("seqinr")
# install.packages("gtools")

library(seqinr)
library(gtools)

phred <- asc( s2c("DDDDCDEDCDDDDBBDDDCC@") ) - 33
phred
```

# Section 3: Mapping RNA-Seq Reads to Genome 

# Section 4: Population Scale Analysis 

## RNA-Seq Population Scale Analysis
Read our results file from RNA-Seq across the ~230 samples (i.e. people)

```{r}
expr <- read.table("rs8067378_ENSG00000172057.6.txt")
```

Look at the data.
```{r}
table(expr$geno)
```

How many samples?
```{r}
nrow(expr)
```

Expression level summary stats
```{r}
summary(expr)
```

```{r}
summary(expr$exp[(expr$geno == "G/G")])
summary(expr$exp[(expr$geno == "A/A")])
summary(expr$exp[(expr$geno == "A/G")])
```

```{r}
# Boxplot
boxplot(exp~geno, data=expr, notch=TRUE)
```

```{r}
library(ggplot2)

ggplot(expr,aes(geno, exp)) + geom_boxplot()
```








































